### ISSUE TITLE
Does not reload code changes when using namespaces

### ISSUE THREAD

**Issue by gururuby** [palkan/action_policy#16]
Hi, I have a controller(Subscriptions) inside the Dashboard module with action manage, eg (app/controllers/dashboard/subscriptions_controller.rb)

```
module Dashboard
  class SubscriptionsController < BaseController
     def manage
        authorize! Subscription.new
     end
  end
end

```
And my policies structure app/policies/dashboard/subscription_policy.rb
```
module Dashboard
  class SubscriptionPolicy < ApplicationPolicy
    def manage?
       true
    end
  end
end
```
When I change the policy(manage?) to false, then the changes do not apply, but everything works fine when I do not use namespaces, and I take subscription_policy to the root of the app / policies directory

Ruby 2.5.1
Rails 5.2.0
Action Policy 0.2.0

**Comment by jacobwalkr** [comment]
Could this be related to #14?

**Comment by gururuby** [comment]
@jacobwalkr may be. There are additions if I call `authorize!` and specify policy directly via the `with` option, changes in the policy are applied without restarting the server

**Comment by ycherniavskyi** [comment]
Got the same reload issue.
After some investigation found the reason - `NamespaceCache` class, which hold Hash with first-time loaded policy class and clear it only on init.

Seams that in development this cash must be disabled or must be clearing in each request.

**Comment by palkan** [comment]
@ycherniavskyi 

Thanks for the investigation!

> Seams that in development this cache must be disabled or must be clearing on each request.

We should add a configuration to disable this cache, and disable it by default in development env.

We can also embrace `app.reload.before_class_unload` callback for Rails 5 and clear the cache.

If anyone is willing to propose a PR – you're welcome! If not – I'll take care of it a little bit later.

**Comment by ycherniavskyi** [comment]
@palkan could you please explain a bit the exact purpose of `NamespaceCache` class. Why namespaced class has the cache but class without namespaced hasn't.

Also as I see from the source, found namespaced policy also cache in the object and in the thread. Will not it be easy and simple just remove `NamespaceCache` mechanism 🙂?

**Comment by palkan** [comment]
> Why namespaced class has the cache but class without namespaced hasn't.

Namespaced lookup is more _expensive_. I've added a [benchmark](https://github.com/palkan/action_policy/blob/master/benchmarks/namespaced_lookup_cache.rb) which shows that it could be 4x slower without cache:

```
Comparison:
               cache:   178745.5 i/s
             cache B:   169435.4 i/s - same-ish: difference falls within error
          no cache A:    95312.7 i/s - same-ish: difference falls within error
          no cache B:    43121.0 i/s - 4.15x  slower
```

> Also as I see from the source, found namespaced policy also cache in the object and in the thread. Will not it be easy and simple just remove NamespaceCache mechanism 🙂?

`NamespaceCache` caches a **policy class** for a given namespace and a policy name, we use it every time we want to initialize a policy within the matching context.
Per-thread and instance caches store **policy instances**, they only re-used within a single request. We still have to resolve a policy class.

**Comment by palkan** [comment]
Fixed by https://github.com/palkan/action_policy/commit/1d4a2e720c0da469d89240abdbd1c44a2daafb1b and released in 0.2.3.
